### PR TITLE
Update device URL to fix 410 error

### PR DESCRIPTION
**Pull request by travipross**
TL;DR: Update URL to fix broken stuff.
-----------

Recently this library (and the dependent Home Assistant integration) appears to have been broken by some changes to the Ayla API. As noted [here](https://github.com/home-assistant/core/issues/44775#issuecomment-1027061596) and [here](https://github.com/home-assistant/core/issues/65367). 

Requests to the old `DEVICE_URL` seem to be returning a 410 error (e.g. when calling `AylaApi.get_devices()`):
```
Traceback (most recent call last):
  File "test.py", line 9, in <module>
    shark_vacs = ayla_api.get_devices()
  File "/home/travipross/git/external/sharkiq/sharkiqpy/ayla_api.py", line 205, in get_devices
    devices = [SharkIqVacuum(self, d) for d in self.list_devices()]
  File "/home/travipross/git/external/sharkiq/sharkiqpy/ayla_api.py", line 192, in list_devices
    devices = resp.json()
  File "/home/travipross/.pyenv/versions/shark/lib/python3.7/site-packages/requests/models.py", line 900, in json
    return complexjson.loads(self.text, **kwargs)
  File "/home/travipross/.pyenv/versions/3.7.7/lib/python3.7/json/__init__.py", line 348, in loads
    return _default_decoder.decode(s)
  File "/home/travipross/.pyenv/versions/3.7.7/lib/python3.7/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/home/travipross/.pyenv/versions/3.7.7/lib/python3.7/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```
 I just set a breakpoint to step through the problem, and I found the [response to the request that fetches the list of devices](https://github.com/ajmarks/sharkiq/blob/c2b12020af287c059b3d4a467d15dcfaa4f14430/sharkiqpy/ayla_api.py#L191) was coming back with `resp.status_code=410` and `resp.text=This resource is not available over http`. Not really sure if the error text is super relevant here, but there's no JSON in that response payload, hence the uncaught exception. 

I honestly can't remember exactly where I got this new URL with complete certainty, but I was able to find it in my terminal history from back when I was messing around with `curl` against the API. 

A while back I had been stepping through the API, hitting the `/devices/<id>/properties` endpoint to see if I could figure out how to extract map data. Well at the time, I think I had spotted the property `GET_Visual_Floor_1`, and noticed it had a `value` field which contained an Ayla API URL. At the time I had thankfully tried to hit that URL (unfortunately without any luck; I could download a file but not make any sense of it), which left it in my shell history.
```json
{
    "property": {
      "type": "Property",
      "name": "GET_Visual_Floor_1",
      "base_type": "file",
      "read_only": true,
      "direction": "output",
      "scope": "user",
      "data_updated_at": "2022-01-03T18:03:34Z",
      "key": 12345678,
      "device_key": 87654321,
      "product_name": "Jaws",
      "track_only_changes": false,
      "display_name": "Visual Floor 1",
      "host_sw_version": false,
      "time_series": false,
      "derived": false,
      "app_type": null,
      "recipe": null,
      "value": "https://ads-apiv1-sharkue1.aylanetworks.com/apiv1/devices/12345678/properties/GET_Visual_Floor_1/datapoints/12341234-12324-1234-1234-123412341234.json",
      "generated_from": null,
      "generated_at": null,
      "denied_roles": [],
      "ack_enabled": false,
      "retention_days": 30
    }
``` 

I didn't realize until today that the subdomain of aylanetworks.com was different than the `DEVICE_URL` from `sharkiq` I had originally used to fetch this property payload. Sure enough, if I use that domain as the `DEVICE_URL` with `sharkiq`, it can successfully list my devices and access the methods as expected.

Perhaps the folks at Shark/Ayla had migrated some device functionality over to a new URL, and I was just lucky enough to catch it as part of a response during some transitional period where both URLs were live, but the new one was being returned in property payloads.

Nonetheless, I believe this should fix the issues over at the [Home Assistant repo](https://github.com/home-assistant/core/issues/65367).

### CODE REVIEW

**Comment by TheOneOgre** [comment]
Based on what I'm seeing, wouldn't the LOGIN_URL variable need to have it's URL changed too?

**Comment by travipross** [comment]
> Based on what I'm seeing, wouldn't the LOGIN_URL variable need to have it's URL changed too?

You would think, but I'm still able to log in against the same `LOGIN_URL` as it is now. Changing that value from `https://user-field.aylanetworks.com` to `https://user-apiv1-sharkue1.aylanetworks.com` to match the `DEVICE_URL` actually doesn't seem to hit anything.

I wish I knew how @ajmarks actually found thes URLs in the first place, so I could be a little more confident in the "why", but this seems to be the answer to the "how", at least from my quick pass at local/manual testing.

**Comment by TheOneOgre** [comment]
> > Based on what I'm seeing, wouldn't the LOGIN_URL variable need to have it's URL changed too?
> 
> You would think, but I'm still able to log in against the same `LOGIN_URL` as it is now. Changing that value from `https://user-field.aylanetworks.com` to `https://user-apiv1-sharkue1.aylanetworks.com` to match the `DEVICE_URL` actually doesn't seem to hit anything.
> 
> I wish I knew how @ajmarks actually found thes URLs in the first place, so I could be a little more confident in the "why", but this seems to be the answer to the "how", at least from my quick pass at local/manual testing.

Sorry, I deleted my original comment shortly after realizing that it was probably an unnecessary change considering you already tested it.

I have done some digging and will continue to do so, looking at the [Ayla API documentation](https://developer.aylanetworks.com/apibrowser/swaggers/DeviceService#!/Devices.json), seems that's where a lot of these URLs were pulled from. Considering they changed the URL and probably changed some code, starting here could be a way to re-reverse engineer access or create a new access method all together if you can access your devices via the original ads-field or ads-dev API link. I'll have to look into it more when I get time.

**Comment by travipross** [comment]
> I have done some digging and will continue to do so, looking at the Ayla API documentation, seems that's where a lot of these URLs were pulled from.

Yes, you're correct about that being where the various endpoints came from, but I assume in order to get the **base URLs** for those endpoints, @ajmarks may have had to do a "Man in the Middle"  exploit and sniff the traffic coming from the official app.

Unfortunately a bit over my head, but maybe I missed something and there's an easier way to find that for sure?

**Comment by TheOneOgre** [comment]
> > I have done some digging and will continue to do so, looking at the Ayla API documentation, seems that's where a lot of these URLs were pulled from.
> 
> Yes, you're correct about that being where the various endpoints came from, but I assume in order to get the **base URLs** for those endpoints, @ajmarks may have had to do a "Man in the Middle" exploit and sniff the traffic coming from the official app.
> 
> Unfortunately a bit over my head, but maybe I missed something and there's an easier way to find that for sure?

Oh yeah I understand what you mean. That was actually my next thought about how to do it. I've got an extra AP I can use specifically for Wireshark sniffing and as long as the messages aren't encrypted, I'll see if I can pull anything from it.

**Comment by USAFPride** [comment]
As this may change again in the future, would it make sense to have the address changeable in the "System Options"?

**Comment by travipross** [comment]
This is not something a user should be in control of modifying under normal circumstances. Currently this library/integration is authenticating against an API which was not intended to be officially supported for 3rd party integration. I suspect there was some light "reverse engineering" done to get this to work in the first place, so I think our best bet is to simply hope that Shark doesn't change it again anytime soon.

But just my $0.02. I just don't think I'd build in support for overriding this URL if we expect (or at least hope) it doesn't change again.

**Comment by TheOneOgre** [comment]
I have to agree with @travipross. 

Having a user changeable URL doesn't make much sense as it's very easy for someone to mess up and misconfigure it, as well as the fact that API changes happen very rarely if ever.

**Comment by creativeosmose** [comment]
@travipross thanks for finding this fix.
Would you by any chance with the equivalent for the europe environment?
(unofficial integration here : https://github.com/Draes/sharkeu)

I tried adding "-eu" at different positions in the domain name, but with no success.

Just to be clear, the unofficial integration works great for Europe, but we encounter the same need for reloading the integration as the US/Worldwide integration.

**Comment by travipross** [comment]
Sorry @creativeosmose, this fix is addressing the fact that the API at the previous `DEVICE_URL` to have changed to a different subdomain.

The daily crashes are not fixed by this change. That will require some reworking of the re-authentication flow, as currently that's the section of the code which raises an uncaught exception.

**Comment by creativeosmose** [comment]
> Sorry @creativeosmose, this fix is addressing the fact that the API at the previous `DEVICE_URL` to have changed to a different subdomain.
> 
> The daily crashes are not fixed by this change. That will require some reworking of the re-authentication flow, as currently that's the section of the code which raises an uncaught exception.

Understood, thanks!

**Review by craig0r — APPROVED** [review]

**Comment by TheOneOgre** [comment]
Awesome the PR has been accepted! Question for anyone who might know, how does HA actually determine when/how to pull data from repos? I've scoured through the source and can't seem to find the code that actually determine the repo to pull the component data from. Is it just baked into the image such as the docker or what?

**Comment by travipross** [comment]
I beleive after this PR gets merged and a new release of `sharkiqpy` is created, [this manifest](https://github.com/home-assistant/core/blob/8d2fac09bb17f825e9447e9ac4817caf75f3148b/homeassistant/components/sharkiq/manifest.json#L6) needs to be updated over in the core project to pull from the newer version.

**Comment by TheOneOgre** [comment]
Interesting, I saw that but didn't see a direct way it chose what repo it came from

What what I understand, changing the domain changes the repo that is pulled from, the requirements changes the branch?(unsure about that) and the code owner changes WHO's repo it's pulled from. 

Just curious how HA actually pulls these things

**Comment by travipross** [comment]
@TheOneOgre The code for the integration itself lives in the `components` directory of the [Home Assistant core repo alongside that manifest file](https://github.com/home-assistant/core/tree/8d2fac09bb17f825e9447e9ac4817caf75f3148b/homeassistant/components/sharkiq). The integration `sharkiq` requires this library (`sharkiqpy`) as a dependency. This repo doesn't contain any of the integration code itself, but the underlying library that the integration uses.

The code in this repo is packaged and published to [PyPi](https://pypi.org/project/sharkiqpy/), which is where Home Assistant looks to for the dependencies listed in the above `manifest.json`.

So first, this PR needs to be made into a versioned release which is published to PyPi. Then the Home Assistant core repo must update the `sharkiq` manifest file so that a newer version of this library is downloaded as a dependency for the integration.

**Comment by TheOneOgre** [comment]
@travipross Excellent explanation! Thanks for the detailed explanation of it.

**Comment by Lucidic15** [comment]
Any idea of the timing of when this will be published?

**Comment by TheLastFrame** [comment]
> Any idea of the timing of when this will be published?

This whole project has been stale for over a year now. I doubt anything (like my PR from half a year ago) will be pushed in the near future...
+I've seen in the HA thread, that the issue resolved itself, so this PR is probably redundant

**Comment by lishan89uc** [comment]
the shark devices are not responding again, can someone with the patch check if it is the url again?

**Comment by creativeosmose** [comment]
@lishan89uc can you give your region please?
On the European environment (with the unofficial fix from a couple of weeks ago) it still works.

**Comment by lishan89uc** [comment]
@creativeosmose Region is US, however apparently someone tested it with the new URL and received same unavailable, so URL is not the issue this time.

**Comment by travipross** [comment]
Closing this PR due to this repo going stale. [This fork](https://github.com/JeffResc/sharkiq) seems to be where more current development will take place.

**Comment by travipross** [comment]
Closing this PR due to this repo going stale. [This fork](https://github.com/JeffResc/sharkiq) seems to be where more current development will take place.
